tests/pebble.rs

```diff
@@ -76,66 +81,26 @@ impl PebbleEnvironment {
+    ) -> Result<Vec<CertificateDer<'static>>, Box<dyn StdError>> {
         info!("testing HTTP-01 challenge");
-
-        // Create an order.
-        let identifier = Identifier::Dns("example.com".to_owned());
-        debug!(?identifier, "creating order");
-        let mut order = account
-            .new_order(&NewOrder {
-                identifiers: &[identifier],
-            })
-            .await?;
-        info!(order_url = order.url(), "created order");
-
-        let authorizations = order.authorizations().await?;
-        let mut challenges = Vec::with_capacity(authorizations.len());
-        let mut names = Vec::with_capacity(authorizations.len());
-
-        // Collect up the relevant challenges, provisioning the expected responses as we go.
-        for authz in &authorizations {
-            match authz.status {
-                AuthorizationStatus::Pending => {}
-                AuthorizationStatus::Valid => continue,
-                _ => unreachable!("unexpected authz state: {:?}", authz.status),
-            }
-
-            let challenge = authz
-                .challenges
-                .iter()
-                .find(|c| c.r#type == ChallengeType::Http01)
-                .ok_or("no http01 challenge found")?;
-
-            let Identifier::Dns(identifier) = &authz.identifier;
-
-            // Provision the HTTP-01 challenge response with the Pebble challenge test server.
-            let key_auth = order.key_authorization(challenge);
-            debug!(
-                token = challenge.token,
-                key_auth = key_auth.as_str(),
-                "provisioning HTTP-01 response",
-            );
-            self.add_http01_response(&challenge.token, key_auth.as_str())
-                .await?;
-
-            challenges.push((identifier, &challenge.url));
-            names.push(identifier.clone());
-        }
-
-        // Tell the CA we have provisioned the response for each challenge.
-        for (_, url) in &challenges {
-            debug!(challenge_url = url, "marking challenge ready");
-            order.set_challenge_ready(url).await?;
-        }
-
-        // Poll until the order is ready.
-        poll_until_ready(&mut order).await?;
-
-        // Issue a certificate for the names and test the chain validates to the issuer root.
-        let cert_chain = self.certificate(&mut order, &names).await?;
-
-        Ok((names, cert_chain))
+        self.complete_order(
+            account,
+            identifiers,
+            ChallengeType::Http01,
+            |_identifier, challenge, key_auth| {
+                let token = challenge.token.clone();
+                Box::pin(async move {
+                    debug!(
+                        token,
+                        key_auth = key_auth.as_str(),
+                        "provisioning HTTP-01 response",
+                    );
+                    self.add_http01_response(&token, key_auth.as_str()).await
+                })
+            },
```

Comment:
Ugh, pretty ugly callback stuff... Maybe an `AuthorizationMethod` trait just for the test?

Reply:
That's much cleaner, thanks! It requires declaring some extra structs but I think those make for a good home for some of the misc `add_xxx_response()` fns that I was tacking on to the pebble env before. :+1: